File: minigit-4/py/refs.py
"""Reference management for Mini Git - HEAD, branches, tags."""

from pathlib import Path
from typing import Optional, List, Tuple

from utils import find_repo_root, get_minigit_dir


def get_head(repo_root: Optional[Path] = None) -> Tuple[str, bool]:
    """
    Get HEAD reference.
    Returns (ref_or_sha, is_symbolic).
    If symbolic (is_symbolic=True), ref_or_sha is like 'refs/heads/main'.
    If detached (is_symbolic=False), ref_or_sha is a SHA.
    """
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    head_path = minigit_dir / 'HEAD'
    if not head_path.exists():
        return '', False

    content = head_path.read_text().strip()
    if content.startswith('ref: '):
        return content[5:], True
    else:
        return content, False


def set_head(ref_or_sha: str, symbolic: bool = True, repo_root: Optional[Path] = None):
    """Set HEAD to a reference or SHA."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    head_path = minigit_dir / 'HEAD'
    if symbolic:
        head_path.write_text(f'ref: {ref_or_sha}\n')
    else:
        head_path.write_text(f'{ref_or_sha}\n')


def get_current_branch(repo_root: Optional[Path] = None) -> Optional[str]:
    """Get the current branch name, or None if HEAD is detached."""
    ref, is_symbolic = get_head(repo_root)
    if is_symbolic and ref.startswith('refs/heads/'):
        return ref[11:]  # Remove 'refs/heads/' prefix
    return None


def get_ref(ref_name: str, repo_root: Optional[Path] = None) -> Optional[str]:
    """Get the SHA a reference points to."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    # Try direct ref path
    ref_path = minigit_dir / ref_name
    if ref_path.exists() and ref_path.is_file():
        return ref_path.read_text().strip()

    # Try refs/heads
    ref_path = minigit_dir / 'refs' / 'heads' / ref_name
    if ref_path.exists():
        return ref_path.read_text().strip()

    # Try refs/tags
    ref_path = minigit_dir / 'refs' / 'tags' / ref_name
    if ref_path.exists():
        return ref_path.read_text().strip()

    return None


def set_ref(ref_name: str, sha: str, repo_root: Optional[Path] = None):
    """Set a reference to a SHA."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    ref_path = minigit_dir / ref_name
    ref_path.parent.mkdir(parents=True, exist_ok=True)
    ref_path.write_text(f'{sha}\n')


def delete_ref(ref_name: str, repo_root: Optional[Path] = None) -> bool:
    """Delete a reference. Returns True if deleted."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    ref_path = minigit_dir / ref_name
    if ref_path.exists():
        ref_path.unlink()
        return True
    return False


def list_branches(repo_root: Optional[Path] = None) -> List[str]:
    """List all branch names."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    heads_dir = minigit_dir / 'refs' / 'heads'
    if not heads_dir.exists():
        return []

    branches = []
    for ref_file in heads_dir.rglob('*'):
        if ref_file.is_file():
            rel = ref_file.relative_to(heads_dir)
            branches.append(str(rel))
    return sorted(branches)


def list_tags(repo_root: Optional[Path] = None) -> List[str]:
    """List all tag names."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    tags_dir = minigit_dir / 'refs' / 'tags'
    if not tags_dir.exists():
        return []

    tags = []
    for ref_file in tags_dir.rglob('*'):
        if ref_file.is_file():
            rel = ref_file.relative_to(tags_dir)
            tags.append(str(rel))
    return sorted(tags)


def resolve_head(repo_root: Optional[Path] = None) -> Optional[str]:
    """Resolve HEAD to a SHA."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    ref, is_symbolic = get_head(repo_root)
    if not ref:
        return None

    if is_symbolic:
        ref_path = minigit_dir / ref
        if ref_path.exists():
            return ref_path.read_text().strip()
        return None
    else:
        return ref


def update_branch(branch_name: str, sha: str, repo_root: Optional[Path] = None):
    """Update a branch to point to a SHA."""
    set_ref(f'refs/heads/{branch_name}', sha, repo_root)


def create_branch(branch_name: str, sha: str, repo_root: Optional[Path] = None) -> bool:
    """Create a new branch. Returns False if branch already exists."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    ref_path = minigit_dir / 'refs' / 'heads' / branch_name
    if ref_path.exists():
        return False

    ref_path.parent.mkdir(parents=True, exist_ok=True)
    ref_path.write_text(f'{sha}\n')
    return True


def delete_branch(branch_name: str, repo_root: Optional[Path] = None) -> bool:
    """Delete a branch. Returns True if deleted."""
    return delete_ref(f'refs/heads/{branch_name}', repo_root)


def rename_branch(old_name: str, new_name: str, repo_root: Optional[Path] = None) -> bool:
    """Rename a branch. Returns True if successful."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    old_path = minigit_dir / 'refs' / 'heads' / old_name
    new_path = minigit_dir / 'refs' / 'heads' / new_name

    if not old_path.exists():
        return False
    if new_path.exists():
        return False

    sha = old_path.read_text().strip()
    new_path.parent.mkdir(parents=True, exist_ok=True)
    new_path.write_text(f'{sha}\n')
    old_path.unlink()

    # Update HEAD if it points to old branch
    ref, is_symbolic = get_head(repo_root)
    if is_symbolic and ref == f'refs/heads/{old_name}':
        set_head(f'refs/heads/{new_name}', symbolic=True, repo_root=repo_root)

    return True


def create_tag(tag_name: str, sha: str, repo_root: Optional[Path] = None) -> bool:
    """Create a lightweight tag. Returns False if tag exists."""
    if repo_root is None:
        repo_root = find_repo_root()
    minigit_dir = get_minigit_dir(repo_root)

    ref_path = minigit_dir / 'refs' / 'tags' / tag_name
    if ref_path.exists():
        return False

    ref_path.parent.mkdir(parents=True, exist_ok=True)
    ref_path.write_text(f'{sha}\n')
    return True


def delete_tag(tag_name: str, repo_root: Optional[Path] = None) -> bool:
    """Delete a tag. Returns True if deleted."""
    return delete_ref(f'refs/tags/{tag_name}', repo_root)


def is_branch_merged(branch_name: str, into_branch: str = 'main',
                     repo_root: Optional[Path] = None) -> bool:
    """Check if a branch is merged into another branch."""
    if repo_root is None:
        repo_root = find_repo_root()

    branch_sha = get_ref(f'refs/heads/{branch_name}', repo_root)
    into_sha = get_ref(f'refs/heads/{into_branch}', repo_root)

    if not branch_sha or not into_sha:
        return False

    # Check if branch_sha is an ancestor of into_sha
    from objects import read_object, Commit

    visited = set()
    to_visit = [into_sha]

    while to_visit:
        sha = to_visit.pop()
        if sha in visited:
            continue
        visited.add(sha)

        if sha == branch_sha:
            return True

        try:
            obj_type, data = read_object(sha, repo_root)
            if obj_type == 'commit':
                commit = Commit.deserialize(data)
                to_visit.extend(commit.parents)
        except:
            pass

    return False

File: minigit-4/ts/src/refs.ts
// Reference management (HEAD, branches, tags)

import * as fs from 'fs';
import * as path from 'path';
import { getHeadPath, getHeadsDir, getTagsDir, getRefsDir, getMinigitDir, ensureDir } from './utils';
import { objectExists, resolveShortSha, readObject, parseCommitContent, parseTagContent } from './objects';

export function readHead(repoRoot: string): string {
  const headPath = getHeadPath(repoRoot);
  return fs.readFileSync(headPath, 'utf-8').trim();
}

export function writeHead(repoRoot: string, content: string): void {
  const headPath = getHeadPath(repoRoot);
  fs.writeFileSync(headPath, content + '\n');
}

export function isDetachedHead(repoRoot: string): boolean {
  const head = readHead(repoRoot);
  return !head.startsWith('ref:');
}

export function getCurrentBranch(repoRoot: string): string | null {
  const head = readHead(repoRoot);
  if (head.startsWith('ref: refs/heads/')) {
    return head.slice('ref: refs/heads/'.length);
  }
  return null;
}

export function getHeadCommit(repoRoot: string): string | null {
  return resolveRef(repoRoot, 'HEAD');
}

export function resolveRef(repoRoot: string, ref: string): string | null {
  // Handle special refs
  if (ref === 'HEAD') {
    const head = readHead(repoRoot);
    if (head.startsWith('ref: ')) {
      return resolveRef(repoRoot, head.slice(5));
    }
    // Detached HEAD - direct SHA
    if (/^[0-9a-f]{40}$/.test(head)) {
      return head;
    }
    return null;
  }

  // Handle refs/heads/xxx or refs/tags/xxx
  if (ref.startsWith('refs/')) {
    const refPath = path.join(getMinigitDir(repoRoot), ref);
    if (fs.existsSync(refPath)) {
      const sha = fs.readFileSync(refPath, 'utf-8').trim();
      // Handle symbolic refs
      if (sha.startsWith('ref: ')) {
        return resolveRef(repoRoot, sha.slice(5));
      }
      return sha;
    }
    return null;
  }

  // Try branches
  const branchPath = path.join(getHeadsDir(repoRoot), ref);
  if (fs.existsSync(branchPath)) {
    return fs.readFileSync(branchPath, 'utf-8').trim();
  }

  // Try tags
  const tagPath = path.join(getTagsDir(repoRoot), ref);
  if (fs.existsSync(tagPath)) {
    const sha = fs.readFileSync(tagPath, 'utf-8').trim();
    // Tag might point to annotated tag object
    if (objectExists(repoRoot, sha)) {
      const { type, content } = readObject(repoRoot, sha);
      if (type === 'tag') {
        const tagInfo = parseTagContent(content);
        return tagInfo.object;
      }
    }
    return sha;
  }

  // Try as SHA
  const resolved = resolveShortSha(repoRoot, ref);
  if (resolved) {
    return resolved;
  }

  return null;
}

export function resolveRevision(repoRoot: string, rev: string): string | null {
  // Handle commit^{tree} syntax
  if (rev.endsWith('^{tree}')) {
    const commitRef = rev.slice(0, -7);
    const commitSha = resolveRevision(repoRoot, commitRef);
    if (!commitSha) return null;

    const { type, content } = readObject(repoRoot, commitSha);
    if (type === 'commit') {
      const commitInfo = parseCommitContent(content);
      return commitInfo.tree;
    }
    return null;
  }

  // Handle HEAD:path or ref:path syntax
  if (rev.includes(':')) {
    const [refPart, pathPart] = rev.split(':');
    const commitSha = resolveRevision(repoRoot, refPart);
    if (!commitSha) return null;

    return resolvePathInCommit(repoRoot, commitSha, pathPart);
  }

  // Handle parent references (^ and ~)
  const parentMatch = rev.match(/^(.+?)(\^+|~(\d+))$/);
  if (parentMatch) {
    const base = parentMatch[1];
    const baseSha = resolveRevision(repoRoot, base);
    if (!baseSha) return null;

    if (parentMatch[2].startsWith('^')) {
      // Each ^ goes to first parent
      let sha = baseSha;
      for (let i = 0; i < parentMatch[2].length; i++) {
        const parent = getParent(repoRoot, sha, 0);
        if (!parent) return null;
        sha = parent;
      }
      return sha;
    } else {
      // ~N goes N parents back
      const count = parseInt(parentMatch[3], 10);
      let sha = baseSha;
      for (let i = 0; i < count; i++) {
        const parent = getParent(repoRoot, sha, 0);
        if (!parent) return null;
        sha = parent;
      }
      return sha;
    }
  }

  // Regular ref resolution
  return resolveRef(repoRoot, rev);
}

function getParent(repoRoot: string, sha: string, index: number): string | null {
  if (!objectExists(repoRoot, sha)) return null;

  const { type, content } = readObject(repoRoot, sha);
  if (type !== 'commit') return null;

  const commitInfo = parseCommitContent(content);
  if (index < commitInfo.parents.length) {
    return commitInfo.parents[index];
  }
  return null;
}

function resolvePathInCommit(repoRoot: string, commitSha: string, filePath: string): string | null {
  const { type, content } = readObject(repoRoot, commitSha);
  if (type !== 'commit') return null;

  const commitInfo = parseCommitContent(content);
  return resolvePathInTree(repoRoot, commitInfo.tree, filePath);
}

function resolvePathInTree(repoRoot: string, treeSha: string, filePath: string): string | null {
  const parts = filePath.split('/').filter((p) => p);
  let currentSha = treeSha;

  for (let i = 0; i < parts.length; i++) {
    const { type, content } = readObject(repoRoot, currentSha);
    if (type !== 'tree') return null;

    const { parseTreeContent } = require('./objects');
    const entries = parseTreeContent(content);
    const entry = entries.find((e: { name: string }) => e.name === parts[i]);

    if (!entry) return null;
    currentSha = entry.sha;
  }

  return currentSha;
}

export function updateBranch(repoRoot: string, branchName: string, sha: string): void {
  const branchPath = path.join(getHeadsDir(repoRoot), branchName);
  ensureDir(path.dirname(branchPath));
  fs.writeFileSync(branchPath, sha + '\n');
}

export function deleteBranch(repoRoot: string, branchName: string): boolean {
  const branchPath = path.join(getHeadsDir(repoRoot), branchName);
  if (fs.existsSync(branchPath)) {
    fs.unlinkSync(branchPath);
    return true;
  }
  return false;
}

export function branchExists(repoRoot: string, branchName: string): boolean {
  const branchPath = path.join(getHeadsDir(repoRoot), branchName);
  return fs.existsSync(branchPath);
}

export function getBranches(repoRoot: string): string[] {
  const headsDir = getHeadsDir(repoRoot);
  if (!fs.existsSync(headsDir)) {
    return [];
  }

  const branches: string[] = [];
  readBranchesRecursive(headsDir, '', branches);
  return branches.sort();
}

function readBranchesRecursive(dir: string, prefix: string, branches: string[]): void {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const name = prefix ? `${prefix}/${entry.name}` : entry.name;
    if (entry.isDirectory()) {
      readBranchesRecursive(path.join(dir, entry.name), name, branches);
    } else if (entry.isFile()) {
      branches.push(name);
    }
  }
}

export function createTag(repoRoot: string, tagName: string, sha: string): void {
  const tagPath = path.join(getTagsDir(repoRoot), tagName);
  ensureDir(path.dirname(tagPath));
  fs.writeFileSync(tagPath, sha + '\n');
}

export function deleteTag(repoRoot: string, tagName: string): boolean {
  const tagPath = path.join(getTagsDir(repoRoot), tagName);
  if (fs.existsSync(tagPath)) {
    fs.unlinkSync(tagPath);
    return true;
  }
  return false;
}

export function tagExists(repoRoot: string, tagName: string): boolean {
  const tagPath = path.join(getTagsDir(repoRoot), tagName);
  return fs.existsSync(tagPath);
}

export function getTags(repoRoot: string): string[] {
  const tagsDir = getTagsDir(repoRoot);
  if (!fs.existsSync(tagsDir)) {
    return [];
  }

  const tags: string[] = [];
  const entries = fs.readdirSync(tagsDir, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.isFile()) {
      tags.push(entry.name);
    }
  }
  return tags.sort();
}

export function updateRef(repoRoot: string, refPath: string, sha: string): void {
  const fullPath = path.join(getMinigitDir(repoRoot), refPath);
  ensureDir(path.dirname(fullPath));
  fs.writeFileSync(fullPath, sha + '\n');
}

export function readRef(repoRoot: string, refPath: string): string | null {
  const fullPath = path.join(getMinigitDir(repoRoot), refPath);
  if (fs.existsSync(fullPath)) {
    return fs.readFileSync(fullPath, 'utf-8').trim();
  }
  return null;
}

export function writeSymbolicRef(repoRoot: string, name: string, target: string): void {
  const refPath = name === 'HEAD' ? getHeadPath(repoRoot) : path.join(getMinigitDir(repoRoot), name);
  fs.writeFileSync(refPath, `ref: ${target}\n`);
}

export function readSymbolicRef(repoRoot: string, name: string): string | null {
  const refPath = name === 'HEAD' ? getHeadPath(repoRoot) : path.join(getMinigitDir(repoRoot), name);
  if (!fs.existsSync(refPath)) {
    return null;
  }

  const content = fs.readFileSync(refPath, 'utf-8').trim();
  if (content.startsWith('ref: ')) {
    return content.slice(5);
  }
  return null;
}
